### ISSUE TITLE
Proposal: Task execution graph

### ISSUE THREAD

**Issue by nat-n** [nat-n/poethepoet#26]
Another model for task composition (in addition to sequence tasks) would be to support declaring dependencies between tasks, such that invoking a task with dependencies would first run all its dependencies (and their dependencies and so on).

Thus an execution DAG of tasks could easily be defined as in the following

### Config proposal A

```toml
[tool.poe.tasks]
test = "pytest src"
version.script = "scripts:get_current_version"

  [tool.poe.tasks.build]
  cmd = "docker build -f Dockerfile -t my_app:$VERSION ."
  deps = { "" = ["test -v"], VERSION = "version" }
```

### Config proposal B

```toml
[tool.poe.tasks]
test = "pytest src"
version.script = "scripts:get_current_version"

  [tool.poe.tasks.build]
  cmd = "docker build -f Dockerfile -t my_app:$VERSION ."
  deps = ["test -v"]
  uses = { VERSION = "version" }
```

Note that in this example scenario the dependency on the test task demonstrates that dep tasks could be passed arguments (or even env vars).

This example also demonstrates the output of one task being captured and made available to the dependant task as an env var hence addressing #21. Since this required silencing (capturing) the output of the dependency, this could allow for multiple tasks to be run in parallel if their dependencies are all satisfied at the same time and neither of them will write to the console (related #24).

This could be particularly powerful if combined with #12 , getting us closer to the power of make.

I've started prototyping this already, but comments are more than welcome.

**Comment by nat-n** [comment]
@kikohs I prototyped a solution and made a beta release: https://pypi.org/project/poethepoet/0.11.0b2/

It hasn't been extensively tested, but I'm keen to get feedback on how well it works for your use case, and whether you encounter any issues.

**Comment by nat-n** [comment]
Experimental (but reasonably complete) task execution graph released in v0.11.0.
